Fix mapper visualization

lewtun opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/giotto-learn/giotto-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
This resolves #127, resolves #121, and resolves #128 

#### What does this implement/fix? Explain your changes.
Major refactoring to streamline the plotting APIs from three to two functions:

* `create_static_network()`
* `create_interactive_network()`

A big advantage from these changes is that the signature for both functions is the same and we now enable interactive mapper in 3D!

#### Any other comments?

<!--
We value all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Thanks for contributing!
-->

Commit:
Refactoring of create_network_2d to fix (#121), (#125), (#127), (#128)

- A suggestion of a template for fixing all the above issues in the static and 2d case.
- Some necessary changes also performed in `utils.visualization.py`
- Mapper Quickstart notebook updated to show new behaviour with a pandas dataframe example

NOTE IMPORTANT: this has broken create_interactive_network, which will need to be refactored separately if these changes are accepted. Also, no change has been made to create_network_3d, pending approval for 2d.

lewtun committed:
Merge branch '0.1.4' into fix_mapper_visualization

lewtun committed:
Rename node_color argument to node_color_statistic

lewtun committed:
Refactor name of is_node_colors_ndadday to is_node_colors_ndarray

lewtun committed:
Extend functionality from 2D visualisation to 3D

- Does not include the hoverlabel

lewtun committed:
Set hoverlabel background color for 3D visualisation

lewtun committed:
Fix linting in visualisation

lewtun committed:
Extend interactive visualisation to include new API

lewtun committed:
Fix linting

lewtun committed:
Add "Color by" annotation to 3D visualisation

lewtun committed:
Move viridis colorscale to top of dropdown list

Since the default colorscale is viridis, it makes sense to show the user this value at the top of the dropdown

lewtun committed:
Refactor 2D and 3D networks into single function

lewtun committed:
Merge plot_options from 2D and 3D networks and remove colorscale dropdown

lewtun committed:
Merge plot and layout options from 2D and 3D plots

lewtun committed:
Change sliders for filter params to bounded boxes

lewtun committed:
Fix layout options for 2D and 3D plots

lewtun committed:
Add docstring for dim argument

lewtun committed:
Refactor case handling for 2D vs 3D plotting into single if statement

lewtun committed:
Refactor 3D if statement to single clause

lewtun committed:
Remove 2D and 3D plotting functions

lewtun committed:
Remove canonical node ID from hover text

lewtun committed:
Update init for static function

lewtun committed:
Add 3D functionality to interactive plot

lewtun committed:
Clone pipeline in static plotting

lewtun committed:
Update quickstart notebook for new API changes

lewtun committed:
Modularise case checking to helper functions

lewtun committed:
Update quickstart notebook for new API

lewtun committed:
Fix bug with layout dim refactoring

lewtun committed:
Remove unused helper function for color scales

lewtun committed:
Sort imports in visualisation.py

lewtun committed:
Add TODO comment on JIT computation for color by columns

lewtun commented:
@ulupo the PR is now ready for your review. 

I've opted to retain the scroll zoom functionality for the static 2D plot over the just-in-time computation of the color by columns dropdown. (Note the scroll zoom is not available in the interactive mode, and never was.)

In the meantime, I'll post a question on the plotly forums to see if it is possible to achieve the best of both worlds - scroll zoom and just-in-time computation.

ulupo commented on giotto/mapper/utils/visualization.py:
```diff
@@ -10,14 +12,13 @@ def _get_node_size(node_elements):
 def _get_node_text(graph):
     return [
-        'Node Id:{}<br>Node size:{}<br>Pullback Id:{}<br>Cluster label:{}'
+        'Node Id:{}<br>Node size:{}<br>Cluster label:{}'
```
Is 'Node Id' a good name? The `for interval_id` makes me think this is the id of the interval in the filter range.

ulupo replied:
'Cluster label' is, AFAIK, a relative label within the pullback cover set in question. Discussion: is the name sufficiently descriptive? Or do we need to be even more explicit, e.g. with 'Relative cluster label' or sth? I really am not sure.

lewtun replied:
Both good points. Since it seems reasonable to assume most of our users will have some basic knowledge of Mapper, why not adopt the following:

* Pullback cover set Id
* Partial cluster label

I propose "partial" because that's how Carlsson et al refer to clustering in their paper.

ulupo replied:
Approved!

ulupo commented on giotto/mapper/utils/visualization.py:
```diff
@@ -27,50 +28,111 @@ def set_node_sizeref(node_elements, node_scale=12):
+    return column_color_buttons
+
+
+def _is_array_or_dataframe(color_variable, data):
```
It seems to me this function's name (and docstring) does not reflect well its behaviour, as it is also determining whether color_variable is a (fit-)transformer, None, or something else.

lewtun replied:
Good point. I will change it to `_infer_color_variable_kind`

ulupo commented on giotto/mapper/utils/visualization.py:
```diff
@@ -27,50 +28,111 @@ def set_node_sizeref(node_elements, node_scale=12):
+        color_variable_kind = 'fit_transformer'
+    elif callable(color_variable):
+        color_variable_kind = 'callable'
+    elif color_variable is None:
```
Now that this function has been isolated in utils, I wonder if the correct string values for the last two cases should be 'none' and 'else' respectively.

ulupo replied:
(Or something similarly generic.)

lewtun replied:
Done as you suggest

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -15,36 +15,56 @@
+        a numpy ndarray or pandas dataframe, it must have the same length as
+        `data` and is interpreted as a quantity of interest according to
+        which each node of the Mapper graph is to be coloured (see
+        :attr:`node_colors`). ``None`` is equivalent to passing `data`. If an
```
Replace `node_colors` with `node_color_statistic`?

lewtun replied:
Good catch!

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -15,36 +15,56 @@
+    layout_dim : int, default: ``2``
+        The number of dimensions for the layout. Can be 2 or 3.
+
+    color_variable : column index or name, or list of such, \
```
Perhaps, to avoid the current long line describing available types, this entry can be rewritten as follows:
- Remove the long list of object types and change it to 'object or None, optional, default: ``None``'
- Detail the various possibilities in a bullet point list (careful with newlines and indents, for reference you can check the docstring for `PairwiseDistance`)

lewtun replied:
Agreed. You'll see my attempt with the next push.

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -15,36 +15,56 @@
+        the Mapper graph, and its values are used directly for node
+        coloring, ignoring `color_variable`. Otherwise, it must be a
+        callable object and is used to obtain a summary statistic within
+        each Mapper node of the quantity specified by :attr:`node_colors`.
```
I think `node_colors` should be `color_variable` here?

lewtun replied:
Yes!

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -15,36 +15,56 @@
+        callable object and is used to obtain a summary statistic within
+        each Mapper node of the quantity specified by :attr:`node_colors`.
+
+    color_by_columns_dropdown : bool, optional, default: ``True``
```
Should the default be False, at the very least to avoid unnecessary computation in the current implementation (note: change both here and in the code)?

lewtun replied:
Agreed.

ulupo replied:
Has this been done?

lewtun replied:
Yes

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -295,153 +172,216 @@ def create_network_3d(graph, data, node_pos, node_color, columns_to_color=None,
-        Key-value pairs (column_name, column_index) to specify which columns of
-        :attr:`data` to color the graph by. Generates a dropdown widget to
-        select the columns to color by.
+    pipeline : :class:`MapperPipeline` object
```
Same comments on dosctring and default values as for static case

lewtun replied:
Done

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -54,238 +74,95 @@ def create_network_2d(graph, data, node_pos, node_color,
-        line=plot_options['edge_trace_line'],
-        hoverinfo=plot_options['edge_trace_hoverinfo'],
-        mode=plot_options['edge_trace_mode'])
+    Examples
```
Create backlog entries for this and remove section for the time being

lewtun replied:
Done

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -54,238 +74,95 @@ def create_network_2d(graph, data, node_pos, node_color,
-    plotly_kwargs : dict, optional, default: ``None``
-        Keyword arguments to configure the Plotly Figure.
+    # Determine whether color_variable is an array or pandas series/dataframe
+    # containing scalar values
```
Add (fit-)transformer and column index or list of indices in comment, or simply remove comment

lewtun replied:
Removed since it's a bit redundant to have a docstring that says the same thing

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -485,30 +425,33 @@ def get_cluster_params_widgets(param, value):
-        return create_network_2d(graph, data, node_pos, node_color,
-                                 columns_to_color, plotly_kwargs)
+        return create_static_network(
+            pipe, data, layout, dim,
```
Technically, the cloning done internally by `create_static_network` is unnecessary here. Proposal: add a 'clone_pipeline' kwarg to `create_static_network`, which defaults to True, and controls whether a clone is made or not? This way we can pass False here, as cloning is already done at the beginning of the call to `create_interactive_network`.

lewtun replied:
Good idea - I'll implement it

lewtun replied:
I've actually opted to remove the `get_figure()` function altogether - it was previously needed to perform some intermediate calculations but with the new API it is redundant

ulupo replied:
I agree, it is just wrapping. But I think the discussion point on cloning remains valid.

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -295,153 +172,216 @@ def create_network_3d(graph, data, node_pos, node_color, columns_to_color=None,
-    # configuration of parameters and re-executes cell in Jupyter notebook
     pipe = clone(pipeline)
 
     def get_cover_params_widgets(param, value):
```
Can we have the cover type here too? I.e. balanced vs uniform

lewtun replied:
Good catch!

ulupo commented on giotto/mapper/visualization.py:
```diff
@@ -485,30 +425,33 @@ def get_cluster_params_widgets(param, value):
-    def update_figure(old_figure, new_figure):
+    def update_figure(old_figure, new_figure, dim):
         # TODO could this be abstracted to node and edge traces and metadata
-        #  information without need for creating a full new figure object
```
For the record: extra blank space was created by PyCharm because of the TODO

lewtun committed:
Change name of hoverlabel text variables

lewtun committed:
Refactor name of _is_array_or_dataframe() utility function

lewtun committed:
Fix docstring to reflect change from node_colors to node_color_statistic

lewtun committed:
Fix name of attribute names in docstrings and use US english for color

lewtun committed:
Add clone_pipeline bool and remove get_figure() utility function

lewtun committed:
Migrate examples docstring to issue #152

lewtun committed:
Refactor setting of cover parameters to be generic for ints and floats

lewtun committed:
Add string-based widget for cover

lewtun committed:
Change color_by_column default value to False

lewtun committed:
Tidy up color_variable explanation in docstrings

lewtun committed:
Update quickstart notebook to match API changes

lewtun committed:
Remove unused get_figure() helper function

lewtun committed:
Remove submit button by setting continuous_update False

lewtun committed:
Clean up commented out code from removal of submit button

lewtun committed:
Add comment about cloning pipeline

lewtun committed:
Refactor update_figure() to accept node and edge traces

Saves memory by avoiding need to create a new plotly Figure object with each update

lewtun committed:
Clean up function signatures and variable names for node and edge traces

lewtun committed:
Tweak whitespace in comment

lewtun committed:
Refactor helper function to reflect changes to return

lewtun committed:
Update quickstart notebook to reflect changes

lewtun committed:
Fix docstring to remove attr

lewtun committed:
Fix docstring for color by columns dropdown default value
